runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -0,0 +1,683 @@
+   _cachedMethodMonitor(TR::Monitor::create("JIT-JITServerAOTCacheCachedMethodMonitor"))
+   {
+   bool allMonitors = _classLoaderMonitor && _classMonitor && _methodMonitor &&
+                      _classChainMonitor && _wellKnownClassesMonitor &&
+                      _aotHeaderMonitor && _cachedMethodMonitor;
+   if (!allMonitors)
+      throw std::bad_alloc();
+   }
+
+JITServerAOTCache::~JITServerAOTCache()
+   {
+   freeMapValues(_classLoaderMap);
+   freeMapValues(_classMap);
+   freeMapValues(_methodMap);
+   freeMapValues(_classChainMap);
+   freeMapValues(_wellKnownClassesMap);
+   freeMapValues(_aotHeaderMap);
+   freeMapValues(_cachedMethodMap);
+
+   TR::Monitor::destroy(_classMonitor);
+   TR::Monitor::destroy(_classLoaderMonitor);
+   TR::Monitor::destroy(_methodMonitor);
+   TR::Monitor::destroy(_classChainMonitor);
+   TR::Monitor::destroy(_wellKnownClassesMonitor);
+   TR::Monitor::destroy(_aotHeaderMonitor);
+   TR::Monitor::destroy(_cachedMethodMonitor);
+   }
+
+
+// Helper macros to make the code for printing class and method names to vlog more concise
+#define RECORD_NAME(record) (int)(record).nameLength(), (const char *)(record).name()
+#define LENGTH_AND_DATA(str) J9UTF8_LENGTH(str), (const char *)J9UTF8_DATA(str)
+#define ROMMETHOD_NAS(romMethod) \
+   LENGTH_AND_DATA(J9ROMMETHOD_NAME(romMethod)), LENGTH_AND_DATA(J9ROMMETHOD_SIGNATURE(romMethod))
+
+
+const AOTCacheClassLoaderRecord *
+JITServerAOTCache::getClassLoaderRecord(const uint8_t *name, size_t nameLength)
+   {
+   TR_ASSERT(nameLength, "Empty class loader identifying name");
+   OMR::CriticalSection cs(_classLoaderMonitor);
+
+   auto it = _classLoaderMap.find({ name, nameLength });
+   if (it != _classLoaderMap.end())
+      return it->second;
+
+   auto record = AOTCacheClassLoaderRecord::create(_classLoaderMap.size() + 1, name, nameLength);
+   addToMap(_classLoaderMap, it, { record->data().name(), record->data().nameLength() }, record);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+         "AOT cache %s: created class loader ID %zu -> %.*s",
+         _name.c_str(), record->data().id(), (int)nameLength, (const char *)name
+      );
+
+   return record;
+   }
+
+const AOTCacheClassRecord *
+JITServerAOTCache::getClassRecord(const AOTCacheClassLoaderRecord *classLoaderRecord, const J9ROMClass *romClass)
+   {
+   JITServerROMClassHash hash;
+   if (auto cache = TR::CompilationInfo::get()->getJITServerSharedROMClassCache())
+      hash = cache->getHash(romClass);
+   else
+      hash = JITServerROMClassHash(romClass);
+
+   OMR::CriticalSection cs(_classMonitor);
+
+   auto it = _classMap.find({ classLoaderRecord, &hash });
+   if (it != _classMap.end())
+      return it->second;
+
+   auto record = AOTCacheClassRecord::create(_classMap.size() + 1, classLoaderRecord, hash, romClass);
```

Comment:
The AOT cache is currently append-only; records are never removed. I'll add a comment explaining that and how new record IDs are allocated.

Reply:
On a second thought, it would be better to maintain a separate "last used ID" and increment it after adding each new record. That way it will be easier to support record deletion (e.g. if we want to implement purging of old unused methods).